ngcsimlib/operations.py

```diff
@@ -0,0 +1,78 @@
+from abc import abstractmethod
+
+class BaseOp():
+    is_compilable = True
+
+    @staticmethod
+    @abstractmethod
+    def operation(*sources):
+        pass
+
+    def __init__(self, *sources):
+        self.sources = sources
+        self.destination = None
+
+    def __call__(self, *args, **kwargs):
+        self.resolve(self.value)
+
+    def set_destination(self, destination):
+        self.destination = destination
+
+    @property
+    def value(self):
+        return self.operation(*[s.value for s in self.sources])
+
+    def resolve(self, value):
+        if self.destination is not None:
+            self.destination.set(value)
+
+    def dump(self):
+        class_name = self.__class__.__name__
+
+        source_array = []
+        for source in self.sources:
+            if isinstance(source, BaseOp):
+                source_array.append(source.dump())
+            else:
+                source_array.append(source.name)
+
+        destination = self.destination.name if self.destination is not None else None
+
+
+        return {"class": class_name, "sources": source_array, "destination": destination}
```

Comment:
inconsistent spacing -- remove space above return and add 1-2 spaces after it (to sep out functions)